Add PowerShell logging tests for MacOS and Linux

dantraMSFT opened the pull request:
This PR fixes the logging issue on Linux where logging is intialized before -settingsFile is parsed causing custom log settings to be ignored. (see ConsoleHost.cs and ManagedEntrance.cs)
Fixes https://github.com/PowerShell/PowerShell/issues/6085

The PR also includes basic logging tests for Linux and MacOS.   PSSyslog.psm1 contains the functions to retrieve selected logged items (based on PowerShell's log id and a timestamp) and Logging.Tests.ps1 contains test for Linux and MacOS.

Fixes https://github.com/PowerShell/PowerShell/issues/5312
Fixes https://github.com/PowerShell/PowerShell/issues/5311

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [X] Issue filed - Issue link: https://github.com/PowerShell/PowerShell/issues/6085
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
    - [x] [Add `[feature]` if the change is significant or affectes feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

dantraMSFT committed:
Test cmdlet for consuming a powershell syslog file.

dantraMSFT committed:
Support -After with TotalCount and Tail
Two basic tests illustrating usage.

dantraMSFT committed:
Adjust $syslogfile and detect unsupported syslog configurations

iSazonov commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,188 @@
+        $powershell = Join-Path -Path $PsHome -ChildPath "pwsh"
+
+        # TODO: Adjust for Linux distro, if needed.
+        $syslogFile = 'var/log/syslog'
```
It seems  `/var/log/system.log` on MacOS.

dantraMSFT replied:
We're using os_log, on MacOS.  I haven't started that testing work yet which is why these tests are constrained to $IsLinux

iSazonov replied:
Please consider #6032 too - we could use another log on Linux too.

dantraMSFT replied:
@iSazonov, I am looking at 6032 but I don't understand what you mean by 'we could use another log on Linux too.'

vramm replied:
In reference to dantraMSFT on 27 Jan "we're using os_log, on MacOS..." you mean you're using os_log_with_type [https://developer.apple.com/documentation/os/os_log_with_type?language=objc] or the pre-defined os_log_(info,debug,error,fault) right?

dantraMSFT replied:
@vramm: The details are in src\libpsl-native\src\nativesyslog.cpp - Native_SysLog().

It uses os_log_* functions (fault, error, debug, and log).

dantraMSFT commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,188 @@
+Describe 'SysLog tests on Linux' -Tag 'CI' {
+
+    BeforeAll
+    {
```
Type; will fix.

iSazonov replied:
The PR #4618

dantraMSFT replied:
@iSazonov, is your reference to PR 4618 a heads up or was there something specific in the thread?

I fixed the type for BeforeAll as well.

dantraMSFT commented on test/tools/Modules/PSSysLog/PSSysLog.psm1:
```diff
@@ -0,0 +1,299 @@
+        return $item
+    }
+}
+function ConvertFrom-SysLog
```
Will do.

dantraMSFT commented on test/tools/Modules/PSSysLog/PSSysLog.psd1:
```diff
@@ -0,0 +1,10 @@
+GUID = '56b63338-045c-4697-a24b-5a756268c8b2'
+Author = 'Dan Travison'
+CompanyName = 'Microsoft'
+Copyright = 'Copyright (c) Microsoft Corporation. All rights reserved.'
```
? Are you saying there should not be a copyright here?  Every other module under test\tools\modules has copyright information.

iSazonov replied:
I haven't full understanding that is a policy because the product is under MIT but many files contains Microsoft copyright. CoreFX is under MIT and has MIT in files. Anyway all the new files we created last months do not contain Microsoft copyright and MIT copyright too. /cc @SteveL-MSFT 
I would prefer to see in our contribution documents a direct policy statement.

SteveL-MSFT replied:
@iSazonov can you open a new issue for Copyrights?  I agree that we've been inconsistent and I can follow up with legal and have a specific statement in our docs

iSazonov replied:
Done #6073.

dantraMSFT commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,188 @@
+Import-Module PSSysLog
+
+<#
+    Define enums that mirror the internal enums used
```
These enums are not needed to parse the log. I define them to ensure the tests are using expected values.

dantraMSFT committed:
Refactor to support early terminating of the pipeline

dantraMSFT committed:
WIP

dantraMSFT committed:
Fix limit check
Remove BuildCommaString with Join

dantraMSFT committed:
Use Select-Object -First for clean cancel pipeline
Remove -Results parameter - not needed due to 'clean' canceling of the pipeline via Select-Object
Udpate tests to capture pipeline output instead of -Results parameter

Commit:
Fix PSChannel on Linux to use a bitmask
handle UseAlwaysOperational and UseAlwaysAnalytic keywords but adding to _keywordFilter if the associated channels were selected in configuration
Remove  previous UseAlwaysAnalytic workaround in SysLogProvider.Log

iSazonov commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,188 @@
+            else
+            {
+                # TODO: Look into journalctl and other variations.
+                Write-Warning -Message 'Unsupported Linux syslog configuration.'
```
As I figured from Dan's comments it's still in development - finally this will work on all systems. So my comment is overdue.

dantraMSFT replied:
If i encounter a system where the logic doesn't apply; I'll change it.

Commit:
Address PR feedback.

dantraMSFT commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,188 @@
+    {
+        return $null
+    }
+    [StringBuilder] $sb = [StringBuilder]::new()
```
Yes. Done.

Commit:
On Unix, delay logging console startup until after command-line parsing
Require Test-IsElevated for logging tests to read syslog
Generate unique logId for each logging test.

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -186,6 +186,12 @@ internal sealed partial class ConsoleHost
+                // On Unix, logging has to be deferred until after command-line parsing
+                // completes to allow overriding logging options.
+                PSEtwLog.LogConsoleStartup();
+#endif
```
😕 Is there no way to unify this for all platforms?

dantraMSFT replied:
Yes it is but I'll leave it open for feedback. I don't want.to change the Windows behavior without a little thought. The non-Windows code has to change.

daxian-dbw replied:
I think we should do the config file overriding through an environment variable until the command-line parsing has been moved earlier.

daxian-dbw replied:
I think we should do the config file overriding through an environment variable until the command-line parsing has been moved earlier. Then this `#if UNIX` block won't be needed.

dantraMSFT replied:
It's still open.  Experimental feature has it as a pending design issue.

vramm replied:
Does it actually make sense to have all this output getting sent to the logging process if someone is running Powershell interactively?

Perhaps a command line option to tell Powershell that it's being run as a non-interactive or batch process and then logging of everything or anything above say "informational" could be sent to "syslog" via the standard "C" syslog functions.

That anything above "informational" option could actually be a configuration added to the system file "syslog.conf" or directory "syslog.d" to limit the output at that point instead of doing it inside Powershell.  I currently do this myself to get rid of all the chatter from Powershell as follows:

    # Only log Powershell message above Informational level
    :programname,startswith,"powershell" stop

Obviously this rule is dropping everything sent to the logger by Powershell.  You probably don't want this but I'm not interested in any of the logging from it at the moment and using complex regular expression logging rules can slow down the logger.

dantraMSFT replied:
@vramm You can control the logging level for PowerShell in powershell.config.json.  Take a look at 'Configuring Logging on non-Windows systems' in [about_logging](https://github.com/PowerShell/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Logging.md)

As far as interactive versus batch/non-interactive, there is no concept in PowerShell from a logging perspective.  However, you can start up PowerShell for your interactive sessions using -settingsFile to accomplish what you are looking for.

Commit:
Mark tests as feature
Update comments
Add 'should' statement to output first unexpected log item for test diagnosability.

dantraMSFT committed:
remove dead code and parameters.
Update cmdlet help

dantraMSFT commented:
Dependent on https://github.com/PowerShell/PowerShell/issues/5645

dantraMSFT committed:
Fix build break

dantraMSFT committed:
Recatagorize as CI to force running in travis-ci

dantraMSFT committed:
Export os_log to a text file
Parse exported os_log content

JamesWTruher commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,191 @@
+        & $powershell -NoProfile -SettingsFile $configFile -Command '$env:PSModulePath | out-null'
+
+        # Get log entries from the last 100 that match our id and are after the time we launched Powershell
+        $items = Get-PSSysLog -Path $Settings.SyslogFile -Id $logId -After $now -Tail 100 -Verbose -TotalCount 3
```
there's an assumption that there aren't other PowerShell sessions executing and logging, is that correct? Would it be better to look for a specific PID or something?

dantraMSFT replied:
The generated logId ([guid]::new) uniquely identifies all log output from the powershell process. I don't need to rely on the pid.

JamesWTruher commented on test/tools/Modules/PSSysLog/PSSysLog.psm1:
```diff
@@ -0,0 +1,797 @@
+    # if the current Powershell session was called with administrator privileges,
+    # the Administrator Group's well-known SID will show up in the Groups for the current identity.
+    # Note that the SID won't show up unless the process is elevated.
+    return (([Security.Principal.WindowsIdentity]::GetCurrent()).Groups -contains "S-1-5-32-544")
```
should this be protected by try/catch?
`try { ... } catch { return $false }`

dantraMSFT replied:
That's a placeholder - WIP

dantraMSFT replied:
@JamesWTruher: that code has been replaced.

Commit:
WIP: Add log persistence cmdlets
Update Test-Elevated for MacOS/Linux
Add Test-MacOS,Test-Linux
Change $After parameter behavior to be on or after.

dantraMSFT committed:
Remove unnecessary OS/Elevated checks when consuming text files.

dantraMSFT committed:
Remove comment out checks

dantraMSFT committed:
First draft of MacOS tests
Rename PSSysLog\Import-OSLog to Export-OsLog
Refactor Set-OsLogPersistence to take a bool
Misc comments updated.

dantraMSFT committed:
Finalize Linux tests

dantraMSFT committed:
MacOS Updates

dantraMSFT committed:
MacOS Updates

dantraMSFT committed:
Finish MacOS test support.
Disable MacOS tests pending solution for persisting powershell logged entries.

dantraMSFT committed:
Change Enable parameter to Enable/Disable switches

dantraMSFT committed:
Use Set-OSLogPersistence Enable/Disable switch parameters

dantraMSFT committed:
Re-enable MacOS tests.
Remove log collect logic  and use log show against the system datastore.
Reformat predicate to avoid embedded double quotes.

dantraMSFT commented on test/tools/Modules/PSSysLog/PSSysLog.psm1:
```diff
@@ -0,0 +1,906 @@
+
+        if ([string]::IsNullOrEmpty($id) -eq $false)
+        {
+            if ($parts[[SysLogIds]::Id].StartsWith($id, [StringComparison]::OrdinalIgnoreCase) -eq $false)
```
That seems self-evident (to me.)

dantraMSFT commented on test/tools/Modules/PSSysLog/PSSysLog.psm1:
```diff
@@ -0,0 +1,906 @@
+
+        MMM dd HH:MM:SS machinename id[PID]: message repeated NNN times: [(commitid:TID:CHANNEL) [EventName] Message]
+        #>
+        [string[]] $parts = $content.Split(' ', 8, [System.StringSplitOptions]::RemoveEmptyEntries)
```
Will add an explicit intent comment.

dantraMSFT replied:
Fixed

dantraMSFT commented on test/tools/Modules/PSSysLog/PSSysLog.psm1:
```diff
@@ -0,0 +1,906 @@
+            return $null
+        }
+
+        if ([string]::IsNullOrEmpty($id) -eq $false)
```
Hmm, is it supported/documented behavior that an empty string returns false here?

BrucePay replied:
Yes.

dantraMSFT replied:
Works for me. Fixed

dantraMSFT replied:
Fixed

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ManagedEntrance.cs:
```diff
@@ -33,16 +33,11 @@ public static int Start(string consoleFilePath, [MarshalAs(UnmanagedType.LPArray
+            // NOTE: On Unix, logging has to be deferred until after command-line parsing
+            // complete. On Windows, deferring the call is not needed.
+            PSEtwLog.LogConsoleStartup();
+#endif
```
Same here. The config file overriding should be done through an environment variable until the command-line parsing is moved to an earlier point.

dantraMSFT replied:
Until we decide what changes we want, I think this is the simpler solution.

dantraMSFT committed:
Logging.Tests.ps1 - Add copyright and update WriteLogSettings
PSSysLog.psm1 - update comments and require MacOS/Elevated for Set-OsLogPersistence

dantraMSFT commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,247 @@
+using namespace System.Text
```
Fixed

dantraMSFT commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,247 @@
+        [LogKeyword[]] $LogKeywords = $null
+    )
+
+    [StringBuilder] $sb = [StringBuilder]::new()
```
Yes, much simpler. Fixed

dantraMSFT committed:
Replace [string]::IsNullOrEmpty($id) with $id

dantraMSFT commented:
@TravisEz13 and @daxian-dbw:  Do you have any additional feedback?

dantraMSFT committed:
PR Feedback

dantraMSFT commented on test/tools/Modules/PSSysLog/PSSysLog.psm1:
```diff
@@ -0,0 +1,912 @@
+
+#region Utilities
+
+function Test-Elevated
```
Fixed

dantraMSFT commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,232 @@
+        Export-PSOsLog -After $after -Verbose | Set-Content -Path $contentFile
+        $items = Get-PSOsLog -Path $contentFile -Id $logId -After $after -TotalCount 3 -Verbose
+
+        $items | Should Not Be $null
```
fixed

dantraMSFT commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,232 @@
+        $items = Get-PSOsLog -Path $contentFile -Id $logId -After $after -TotalCount 3 -Verbose
+
+        $items | Should Not Be $null
+        $items.Length | Should BeGreaterThan 1
```
Fixed

dantraMSFT commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,232 @@
+        $items | Should Not Be $null
+        $items.Length | Should BeGreaterThan 1
+        $items[0].EventId | Should Be 'Perftrack_ConsoleStartupStart:PowershellConsoleStartup.WinStart.Informational'
+        $items[1].EventId | Should Be 'Perftrack_ConsoleStartupStop:PowershellConsoleStartup.WinStop.Informational'
```
Fixed

dantraMSFT commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,232 @@
+        # by default, powershell startup should only log informational events are logged.
+        # With Level = Warning, nothing should
+        $items = Get-PSOsLog -Path $contentFile -Id $logId -After $after -TotalCount 3
+        $items | Should Be $null
```
Fixed

dantraMSFT commented:
@TravisEz13, @daxian-dbw  ping.

daxian-dbw commented:
I don't have additional feedback, but I didn't review the whole changes. So I will leave it to @iSazonov @TravisEz13 and @JamesWTruher to sign off the PR.

iSazonov commented on test/tools/Modules/PSSysLog/PSSysLog.psd1:
```diff
@@ -0,0 +1,10 @@
+@{
```
Should we add copyright header?

dantraMSFT replied:
none of the manifest files in the repo have copyright headers; they all have Copyright=""

iSazonov reviewed (commented):
Some minor comments.

dantraMSFT commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,232 @@
+        # Get log entries from the last 100 that match our id and are after the time we launched Powershell
+        $items = Get-PSSysLog -Path $SyslogFile -Id $logId -Tail 100 -Verbose -TotalCount 3
+
+        $items | Should -Not -Be $null
```
Log entries and it's collection are not not strings and I don't think it makes sense to comment that fact that it is not a string.

dantraMSFT commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,232 @@
+        if ($items.Length -gt 2)
+        {
+            # Force reporting of the first unexpected item to help diagnosis
+            $items[2] | Should -Be $null
```
BeNullOrEmpty is for strings; Log entries and it's collection are not not strings and I don't think it makes sense to comment that fact that it is not a string.

dantraMSFT committed:
PR Feedback

dantraMSFT commented on src/System.Management.Automation/utils/tracing/PSEtwLog.cs:
```diff
@@ -30,6 +30,21 @@ static PSEtwLog()
 
+        internal static void LogConsoleStartup()
+        {
+            // Set ETW activity Id
```
Fixed

dantraMSFT commented on test/powershell/Host/Logging.Tests.ps1:
```diff
@@ -0,0 +1,232 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.using namespace System.Text
```
Fixed
